torch/csrc/autograd/engine.cpp

```diff
@@ -192,6 +196,7 @@ auto ReadyQueue::push(FunctionTask item) -> void {
 }
 
 auto ReadyQueue::pop() -> FunctionTask {
+  // Lock for condition variable non_empty_
```

Comment:
So, one thing to clarify here, is that you don't actually need to take out a lock to post a notify on a condition variable; the condition variable is itself a synchronization mechanism. Really what is going on here is that you must take out the lock to check if `heap_.empty()`, and `wait` needs to know to release the lock while it's waiting on the condition variable (so you don't accidentally deadlock.)